File: 673-2/E.cpp
#include <vector>
#include <cstdio>
#include <utility>
using namespace std;

#define pvii pair<vector<int>, int>
#define vi vector<int>

int n, ai, sz, x, minv, binv;
vector<int> a, b;

pair<vector<int>, int> find_inversions(vector<int> arr) {
    if (arr.size() == 1) {
        return pvii(arr, 0);
    }

    vi left_slice = vi(arr.size()/2);
    copy(arr.begin(), arr.begin() + arr.size() / 2, left_slice.begin());
    vi right_slice = vi(arr.size() - arr.size()/2);
    copy(arr.begin() + arr.size() / 2, arr.end(), right_slice.begin());
    
    pvii left = find_inversions(left_slice);
    pvii right = find_inversions(right_slice);

    vi combine = vi();
    int inversions = left.second + right.second;

    while (left.first.size() > 0 && right.first.size() > 0) {
        if (left.first[0] <= right.first[0]) {
            combine.emplace_back(left.first[0]);
            left.first.erase(left.first.begin());
        } else {
            inversions += left.first.size();
            combine.emplace_back(right.first[0]);
            right.first.erase(right.first.begin());
        }
    }

    if (left.first.size() > 0) {
        combine.insert(combine.end(), left.first.begin(), left.first.end());
    } else {
        combine.insert(combine.end(), right.first.begin(), right.first.end());
    }

    return pvii(combine, inversions);
}

int main() {
    scanf("%d", &n);
    for (int i = 0; i < n; ++i) {
        scanf("%d", &ai);
        a.emplace_back(ai);
    }
    sz = 30;
    x = 0;

    minv = find_inversions(a).second;
    for (int j = sz; j >= 0; --j) {
        b.clear();
        for (int i = 0; i < a.size(); ++i) {
            b.emplace_back(a[i] ^ (x + (1<<j)));
        }
        binv = find_inversions(b).second;

        if (binv < minv) {
            x += (1<<j);
            minv = binv;
        }
    }

    printf("%d %d", minv, x);
}

File: 673-2/E.py
import math

t = int(input())
a = list(map(int, input().split()))
sz = 30
x = 0

def find_inversions(arr):
    if len(arr) == 1:
        return (arr, 0)
    
    left = find_inversions(arr[:len(arr)//2])
    right = find_inversions(arr[len(arr)//2:])
    leftarr = left[0]
    rightarr = right[0]
    combine = []
    inversions = left[1] + right[1]
    while len(leftarr) > 0 and len(rightarr) > 0:
        if leftarr[0] <= rightarr[0]:
            combine.append(leftarr.pop(0))
        else:
            inversions += len(leftarr)
            combine.append(rightarr.pop(0))
    if len(leftarr) > 0:
        combine += leftarr
    else:
        combine += rightarr
    return (combine, inversions)

minv = find_inversions(a)[1]
for j in range(sz, -1, -1):
    B = [a[i] ^ (x + (1<<j)) for i in range(len(a))]
    binv = find_inversions(B)

    if binv[1] < minv:
        x += (1 << j)
        minv = binv[1]

print(minv, x)
